File: go/customers.go
// this file is @generated
package meteroid

import (
	"context"
	"net/http"
)

// CustomersListCustomersOptions carries the query and header parameters of
// Customers.ListCustomers.
//
// Optional parameters are pointers; leave them nil to omit them.
type CustomersListCustomersOptions struct {
	// Sort order. Format: `column.direction`. Allowed columns: `name`, `email`, `alias`, `created_at`. Direction: `asc` or `desc`. Default: `created_at.desc`.
	OrderBy *string

	// Page number (0-indexed)
	Page *int32

	// Number of items per page
	PerPage *int32

	Search *string

	Archived *bool
}

// Customers groups the customers operations of the Meteroid API.
type Customers struct {
	client *Client
}

// List customers with optional pagination and search filtering.
func (a *Customers) ListCustomers(ctx context.Context, options *CustomersListCustomersOptions) (*CustomerListResponse, error) {
	req := newRequest(http.MethodGet, "/api/v1/customers")

	if options != nil {
		if options.OrderBy != nil {
			req.SetQueryParam("order_by", *options.OrderBy)
		}
		if options.Page != nil {
			req.SetQueryParam("page", formatInt(int64(*options.Page)))
		}
		if options.PerPage != nil {
			req.SetQueryParam("per_page", formatInt(int64(*options.PerPage)))
		}
		if options.Search != nil {
			req.SetQueryParam("search", *options.Search)
		}
		if options.Archived != nil {
			req.SetQueryParam("archived", formatBool(*options.Archived))
		}
	}

	var out CustomerListResponse
	if err := a.client.execute(ctx, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (a *Customers) CreateCustomer(ctx context.Context, customerCreateRequest CustomerCreateRequest) (*Customer, error) {
	req := newRequest(http.MethodPost, "/api/v1/customers")

	req.SetJSONBody(customerCreateRequest)

	var out Customer
	if err := a.client.execute(ctx, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Retrieve a single customer by ID or alias.
func (a *Customers) GetCustomer(ctx context.Context, idOrAlias string) (*Customer, error) {
	req := newRequest(http.MethodGet, "/api/v1/customers/{id_or_alias}")
	req.SetPathParam("id_or_alias", idOrAlias)

	var out Customer
	if err := a.client.execute(ctx, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (a *Customers) UpdateCustomer(ctx context.Context, idOrAlias string, customerUpdateRequest CustomerUpdateRequest) (*Customer, error) {
	req := newRequest(http.MethodPut, "/api/v1/customers/{id_or_alias}")
	req.SetPathParam("id_or_alias", idOrAlias)

	req.SetJSONBody(customerUpdateRequest)

	var out Customer
	if err := a.client.execute(ctx, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// No linked entity will be deleted. You need to terminate all active subscriptions before archiving a customer, or the call will fail.
func (a *Customers) ArchiveCustomer(ctx context.Context, idOrAlias string) error {
	req := newRequest(http.MethodDelete, "/api/v1/customers/{id_or_alias}")
	req.SetPathParam("id_or_alias", idOrAlias)

	return a.client.execute(ctx, req, nil)
}

// Partially update a customer. Only provided fields will be updated.
func (a *Customers) PatchCustomer(ctx context.Context, idOrAlias string, customerPatchRequest CustomerPatchRequest) (*Customer, error) {
	req := newRequest(http.MethodPatch, "/api/v1/customers/{id_or_alias}")
	req.SetPathParam("id_or_alias", idOrAlias)

	req.SetJSONBody(customerPatchRequest)

	var out Customer
	if err := a.client.execute(ctx, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (a *Customers) GetEffectiveEntitlements(ctx context.Context, idOrAlias string) (*EffectiveEntitlementListResponse, error) {
	req := newRequest(http.MethodGet, "/api/v1/customers/{id_or_alias}/entitlements")
	req.SetPathParam("id_or_alias", idOrAlias)

	var out EffectiveEntitlementListResponse
	if err := a.client.execute(ctx, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Generates a JWT token that grants access to the customer portal.
// The token can be used to access invoices, payment methods, and other portal features.
func (a *Customers) CreatePortalToken(ctx context.Context, idOrAlias string, customerPortalTokenRequest CustomerPortalTokenRequest) (*CustomerPortalTokenResponse, error) {
	req := newRequest(http.MethodPost, "/api/v1/customers/{id_or_alias}/portal-token")
	req.SetPathParam("id_or_alias", idOrAlias)

	req.SetJSONBody(customerPortalTokenRequest)

	var out CustomerPortalTokenResponse
	if err := a.client.execute(ctx, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (a *Customers) UnarchiveCustomer(ctx context.Context, idOrAlias string) error {
	req := newRequest(http.MethodPost, "/api/v1/customers/{id_or_alias}/unarchive")
	req.SetPathParam("id_or_alias", idOrAlias)

	return a.client.execute(ctx, req, nil)
}

File: python/meteroid/api/customers.py
# this file is @generated
import typing as t

from ..models import (
    Customer,
    CustomerCreateRequest,
    CustomerListResponse,
    CustomerPatchRequest,
    CustomerPortalTokenRequest,
    CustomerPortalTokenResponse,
    CustomerUpdateRequest,
    EffectiveEntitlementListResponse,
)
from .common import ApiBaseAsync, ApiBaseSync, decode_response, serialize_query_params


class CustomersAsync(ApiBaseAsync):
    """customers API."""

    async def list_customers(
        self,
        *,
        order_by: t.Optional[str] = None,
        page: t.Optional[int] = None,
        per_page: t.Optional[int] = None,
        search: t.Optional[str] = None,
        archived: t.Optional[bool] = None,
    ) -> CustomerListResponse:
        """List customers with optional pagination and search filtering.

        :param order_by: Sort order. Format: `column.direction`. Allowed columns: `name`, `email`, `alias`, `created_at`. Direction: `asc` or `desc`. Default: `created_at.desc`.
        :param page: Page number (0-indexed)
        :param per_page: Number of items per page"""
        response = await self._request_asyncio(
            method="get",
            path="/api/v1/customers",
            query_params=serialize_query_params(
                {
                    "order_by": order_by,
                    "page": page,
                    "per_page": per_page,
                    "search": search,
                    "archived": archived,
                },
            ),
        )
        return decode_response(response, CustomerListResponse)

    async def create_customer(
        self,
        customer_create_request: CustomerCreateRequest,
    ) -> Customer:
        response = await self._request_asyncio(
            method="post",
            path="/api/v1/customers",
            json_body=customer_create_request.to_dict(),
        )
        return decode_response(response, Customer)

    async def get_customer(
        self,
        id_or_alias: str,
    ) -> Customer:
        """Retrieve a single customer by ID or alias."""
        response = await self._request_asyncio(
            method="get",
            path="/api/v1/customers/{id_or_alias}",
            path_params={
                "id_or_alias": id_or_alias,
            },
        )
        return decode_response(response, Customer)

    async def update_customer(
        self,
        id_or_alias: str,
        customer_update_request: CustomerUpdateRequest,
    ) -> Customer:
        response = await self._request_asyncio(
            method="put",
            path="/api/v1/customers/{id_or_alias}",
            path_params={
                "id_or_alias": id_or_alias,
            },
            json_body=customer_update_request.to_dict(),
        )
        return decode_response(response, Customer)

    async def archive_customer(
        self,
        id_or_alias: str,
    ) -> None:
        """No linked entity will be deleted. You need to terminate all active subscriptions before archiving a customer, or the call will fail."""
        await self._request_asyncio(
            method="delete",
            path="/api/v1/customers/{id_or_alias}",
            path_params={
                "id_or_alias": id_or_alias,
            },
        )

    async def patch_customer(
        self,
        id_or_alias: str,
        customer_patch_request: CustomerPatchRequest,
    ) -> Customer:
        """Partially update a customer. Only provided fields will be updated."""
        response = await self._request_asyncio(
            method="patch",
            path="/api/v1/customers/{id_or_alias}",
            path_params={
                "id_or_alias": id_or_alias,
            },
            json_body=customer_patch_request.to_dict(),
        )
        return decode_response(response, Customer)

    async def get_effective_entitlements(
        self,
        id_or_alias: str,
    ) -> EffectiveEntitlementListResponse:
        response = await self._request_asyncio(
            method="get",
            path="/api/v1/customers/{id_or_alias}/entitlements",
            path_params={
                "id_or_alias": id_or_alias,
            },
        )
        return decode_response(response, EffectiveEntitlementListResponse)

    async def create_portal_token(
        self,
        id_or_alias: str,
        customer_portal_token_request: CustomerPortalTokenRequest,
    ) -> CustomerPortalTokenResponse:
        """Generates a JWT token that grants access to the customer portal.
        The token can be used to access invoices, payment methods, and other portal features."""
        response = await self._request_asyncio(
            method="post",
            path="/api/v1/customers/{id_or_alias}/portal-token",
            path_params={
                "id_or_alias": id_or_alias,
            },
            json_body=customer_portal_token_request.to_dict(),
        )
        return decode_response(response, CustomerPortalTokenResponse)

    async def unarchive_customer(
        self,
        id_or_alias: str,
    ) -> None:
        await self._request_asyncio(
            method="post",
            path="/api/v1/customers/{id_or_alias}/unarchive",
            path_params={
                "id_or_alias": id_or_alias,
            },
        )


class Customers(ApiBaseSync):
    """customers API."""

    def list_customers(
        self,
        *,
        order_by: t.Optional[str] = None,
        page: t.Optional[int] = None,
        per_page: t.Optional[int] = None,
        search: t.Optional[str] = None,
        archived: t.Optional[bool] = None,
    ) -> CustomerListResponse:
        """List customers with optional pagination and search filtering.

        :param order_by: Sort order. Format: `column.direction`. Allowed columns: `name`, `email`, `alias`, `created_at`. Direction: `asc` or `desc`. Default: `created_at.desc`.
        :param page: Page number (0-indexed)
        :param per_page: Number of items per page"""
        response = self._request_sync(
            method="get",
            path="/api/v1/customers",
            query_params=serialize_query_params(
                {
                    "order_by": order_by,
                    "page": page,
                    "per_page": per_page,
                    "search": search,
                    "archived": archived,
                },
            ),
        )
        return decode_response(response, CustomerListResponse)

    def create_customer(
        self,
        customer_create_request: CustomerCreateRequest,
    ) -> Customer:
        response = self._request_sync(
            method="post",
            path="/api/v1/customers",
            json_body=customer_create_request.to_dict(),
        )
        return decode_response(response, Customer)

    def get_customer(
        self,
        id_or_alias: str,
    ) -> Customer:
        """Retrieve a single customer by ID or alias."""
        response = self._request_sync(
            method="get",
            path="/api/v1/customers/{id_or_alias}",
            path_params={
                "id_or_alias": id_or_alias,
            },
        )
        return decode_response(response, Customer)

    def update_customer(
        self,
        id_or_alias: str,
        customer_update_request: CustomerUpdateRequest,
    ) -> Customer:
        response = self._request_sync(
            method="put",
            path="/api/v1/customers/{id_or_alias}",
            path_params={
                "id_or_alias": id_or_alias,
            },
            json_body=customer_update_request.to_dict(),
        )
        return decode_response(response, Customer)

    def archive_customer(
        self,
        id_or_alias: str,
    ) -> None:
        """No linked entity will be deleted. You need to terminate all active subscriptions before archiving a customer, or the call will fail."""
        self._request_sync(
            method="delete",
            path="/api/v1/customers/{id_or_alias}",
            path_params={
                "id_or_alias": id_or_alias,
            },
        )

    def patch_customer(
        self,
        id_or_alias: str,
        customer_patch_request: CustomerPatchRequest,
    ) -> Customer:
        """Partially update a customer. Only provided fields will be updated."""
        response = self._request_sync(
            method="patch",
            path="/api/v1/customers/{id_or_alias}",
            path_params={
                "id_or_alias": id_or_alias,
            },
            json_body=customer_patch_request.to_dict(),
        )
        return decode_response(response, Customer)

    def get_effective_entitlements(
        self,
        id_or_alias: str,
    ) -> EffectiveEntitlementListResponse:
        response = self._request_sync(
            method="get",
            path="/api/v1/customers/{id_or_alias}/entitlements",
            path_params={
                "id_or_alias": id_or_alias,
            },
        )
        return decode_response(response, EffectiveEntitlementListResponse)

    def create_portal_token(
        self,
        id_or_alias: str,
        customer_portal_token_request: CustomerPortalTokenRequest,
    ) -> CustomerPortalTokenResponse:
        """Generates a JWT token that grants access to the customer portal.
        The token can be used to access invoices, payment methods, and other portal features."""
        response = self._request_sync(
            method="post",
            path="/api/v1/customers/{id_or_alias}/portal-token",
            path_params={
                "id_or_alias": id_or_alias,
            },
            json_body=customer_portal_token_request.to_dict(),
        )
        return decode_response(response, CustomerPortalTokenResponse)

    def unarchive_customer(
        self,
        id_or_alias: str,
    ) -> None:
        self._request_sync(
            method="post",
            path="/api/v1/customers/{id_or_alias}/unarchive",
            path_params={
                "id_or_alias": id_or_alias,
            },
        )
